### PR TITLE
Extend `explicit_iter_loop` and `explicit_into_iter_loop`

### LINKED ISSUE #1518: explicit_iter_loop should warn &[T].iter()
```rust
fn main() {
    let s = &[1, 2, 3, 4];
    for a in s.iter() { // .iter() is not necessary
        println!("{:?}", a);
    }
}
```

### PR DESCRIPTION
**Pull request by Jarcho**
fixes #1518

Some included cleanups
* Split `for_loop` test into different files for each lint (partially).
* Move handling of some `into_iter` cases from `explicit_into_iter`.

---

changelog: Enhancement: [`explicit_iter_loop`]: Now also handles types that implement `IntoIterator`.
[#10416](https://github.com/rust-lang/rust-clippy/pull/10416)

changelog: Sugg: [`explicit_into_iter_loop`]: The suggestion now works on mutable references.
[#10416](https://github.com/rust-lang/rust-clippy/pull/10416)

### CODE REVIEW

**Comment by Jarcho** [comment]
ping @Manishearth

File: clippy_lints/src/loops/explicit_into_iter_loop.rs [hunk 1]
```diff
@@ -5,15 +5,76 @@ use clippy_utils::source::snippet_with_applicability;
 use rustc_errors::Applicability;
 use rustc_hir::Expr;
 use rustc_lint::LateContext;
+use rustc_middle::ty::adjustment::{Adjust, Adjustment, AutoBorrow, AutoBorrowMutability};
 use rustc_span::symbol::sym;
 
+#[derive(Clone, Copy)]
+enum AdjustKind {
```
Comment (Manishearth) [inline clippy_lints/src/loops/explicit_into_iter_loop.rs:12]: sugestion: would it be possible to make these adjustkind APIs generally useful in utils?

**Review by Manishearth — APPROVED** [review]
overall lgtm, nice improvement

r=me whether or not you decide to move stuff into utils

File: clippy_lints/src/loops/explicit_into_iter_loop.rs (hunk 1 above)
Comment (Jarcho) [reply clippy_lints/src/loops/explicit_into_iter_loop.rs:12]: I'm not sure how much can be pulled out. Every time I've had to handle adjustments there's always been some weird cases to handle.

**Pull request merged**